python/cudf_polars/cudf_polars/dsl/utils/aggregations.py

```diff
@@ -0,0 +1,245 @@
+            # (potentially masked) child. This is safe because we recursed
+            # to ensure there are no nested aggregations.
+            return (
+                [expr.NamedExpr(name, agg.reconstruct([child]))],
+                expr.NamedExpr(name, expr.Col(agg.dtype, name)),
+                True,
+            )
+        elif agg.name == "sum" and is_top:
+            # In polars sum(empty_group) => 0, but in libcudf sum(empty_group) => null
+            # So must post-process by replacing nulls, but only if we're a "top-level" agg.
+            rep = expr.Literal(
+                agg.dtype, pa.scalar(0, type=plc.interop.to_arrow(agg.dtype))
+            )
+            return (
+                [named_expr],
+                expr.NamedExpr(
+                    name,
+                    expr.UnaryFunction(
+                        agg.dtype, "fill_null", (), expr.Col(agg.dtype, name), rep
+                    ),
+                ),
+                True,
+            )
+        else:
+            return [named_expr], expr.NamedExpr(name, expr.Col(agg.dtype, name)), True
+    if agg.is_pointwise:
+        aggs, posts, has_aggs = _decompose_aggs(
+            (expr.NamedExpr(next(name_generator), child) for child in agg.children),
+            name_generator,
+            is_top=False,
+        )
+        if any(has_aggs):
+            # Any pointwise expression can be handled either by
+            # post-evaluation (if outside an aggregation).
+            return (
+                aggs,
+                expr.NamedExpr(name, agg.reconstruct([p.value for p in posts])),
+                True,
+            )
+        else:
+            # Or pre-evaluation if inside an aggregation.
+            return [named_expr], expr.NamedExpr(name, expr.Col(agg.dtype, name)), False
+    raise NotImplementedError(f"No support for {type(agg)} in groupby")
+
+
+def _decompose_aggs(
+    aggs: Iterable[expr.NamedExpr],
+    name_generator: Generator[str, None, None],
+    *,
+    is_top: bool,
+) -> tuple[list[expr.NamedExpr], Sequence[expr.NamedExpr], Sequence[bool]]:
+    new_aggs, post, has_aggs = zip(
+        *(decompose_single_agg(agg, name_generator, is_top=is_top) for agg in aggs),
+        strict=True,
+    )
+    return (
+        list(itertools.chain.from_iterable(new_aggs)),
+        post,
+        has_aggs,
+    )
+
+
+def decompose_aggs(
+    aggs: Iterable[expr.NamedExpr], name_generator: Generator[str, None, None]
+) -> tuple[list[expr.NamedExpr], Sequence[expr.NamedExpr]]:
+    """
+    Process arbitrary aggregations into a form we can handle in grouped aggregations.
+
+    Parameters
+    ----------
+    aggs
+        List of aggregation expressions
+    name_generator
+        Generator of unique names for temporaries introduced during decomposition.
+
+    Returns
+    -------
+    tuple
+        Three-tuple of expressions to evaluate before the aggregation
+        (as a ``Select`), aggregation expressions for the ``GroupBy``
```

Comment:
```suggestion
        (as a ``Select``), aggregation expressions for the ``GroupBy``
```

Reply:
And maybe more of the description is off since this seems to be a two-tuple now.

Reply:
Fixed, thanks.